instrumentation-genai/opentelemetry-instrumentation-cohere-v2/src/opentelemetry/instrumentation/cohere_v2/utils.py

```diff
@@ -0,0 +1,199 @@
+    tool_calls: Optional[List[cohere.ToolCallV2]] = get_property_value(item, "tool_calls")
+    if tool_calls is None:
+        return None
+
+    calls = []
+    for tool_call in tool_calls:
+        tool_call_dict = {}
+        call_id = get_property_value(tool_call, "id")
+        if call_id:
+            tool_call_dict["id"] = call_id
+
+        tool_type = get_property_value(tool_call, "type")
+        if tool_type:
+            tool_call_dict["type"] = tool_type
+
+        func = get_property_value(tool_call, "function")
+        if func:
+            tool_call_dict["function"] = {}
+
+            name = get_property_value(func, "name")
+            if name:
+                tool_call_dict["function"]["name"] = name
+
+            arguments = get_property_value(func, "arguments")
+            if capture_content and arguments:
+                if isinstance(arguments, str):
+                    arguments = arguments.replace("\n", "")
+                tool_call_dict["function"]["arguments"] = arguments
+
+        calls.append(tool_call_dict)
+    return calls
+
+
+def get_property_value(obj, property_name):
+    if isinstance(obj, dict):
+        return obj.get(property_name, None)
+
+    return getattr(obj, property_name, None)
+
+
+def set_server_address_and_port(client_instance: cohere.client_v2.V2Client, attributes):
+    base_client = getattr(client_instance, "_client_wrapper", None)
+    base_url = getattr(base_client, "_base_url", None)
+    if not base_url:
+        return
+
+    port = -1
+    url = urlparse(base_url)
+    attributes[ServerAttributes.SERVER_ADDRESS] = url.hostname
+    port = url.port
+
+    if port and port != 443 and port > 0:
+        attributes[ServerAttributes.SERVER_PORT] = port
+
+
+def get_llm_request_attributes(
+    kwargs,
+    client_instance: cohere.client_v2.V2Client,
+    operation_name=GenAIAttributes.GenAiOperationNameValues.CHAT.value,
+):
+    attributes = {
+        GenAIAttributes.GEN_AI_OPERATION_NAME: operation_name,
+        GenAIAttributes.GEN_AI_SYSTEM: GenAIAttributes.GenAiSystemValues.COHERE.value,
+        GenAIAttributes.GEN_AI_REQUEST_MODEL: kwargs.get("model"),
+        GenAIAttributes.GEN_AI_REQUEST_MAX_TOKENS: kwargs.get("max_tokens"),
+        GenAIAttributes.GEN_AI_REQUEST_FREQUENCY_PENALTY: kwargs.get(
+            "stop_sequences"
+        ),
+        GenAIAttributes.GEN_AI_REQUEST_TEMPERATURE: kwargs.get("temperature"),
+        # TODO: Add to sem conv
+        "gen_ai.cohere.request.seed": kwargs.get("seed"),
+        GenAIAttributes.GEN_AI_REQUEST_PRESENCE_PENALTY: kwargs.get(
+            "presence_penalty"
+        ),
+        GenAIAttributes.GEN_AI_REQUEST_FREQUENCY_PENALTY: kwargs.get(
+            "frequency_penalty"
+        ),
+        GenAIAttributes.GEN_AI_REQUEST_TOP_K: kwargs.get("k"),
+        GenAIAttributes.GEN_AI_REQUEST_TOP_P: kwargs.get("p"),
+    }
+    response_format = kwargs.get("response_format")
+    if response_format:
+        # TODO: Add to sem conv
```

Comment:
Should we make the openai generic instead?

Reply:
so you mean remove the openai one and make a generic one for this entry: https://github.com/open-telemetry/semantic-conventions/blob/abd92c153b627efd3222e6b46828c5e94bf1b2cd/model/gen-ai/spans.yaml#L88

I think it can make sense, and in any case we can open an issue or PR about any cohere specific attributes, such as these. That PR could be closed by a normalization.

@karthikscale3 you happen to know how many other LLM providers have a concept of response_format?

Reply:
> so you mean remove the openai one and make a generic one for this entry: https://github.com/open-telemetry/semantic-conventions/blob/abd92c153b627efd3222e6b46828c5e94bf1b2cd/model/gen-ai/spans.yaml#L88

Yes

Reply:
@lmolkova Is it possible to have a generic attribute but have each implementation (openai, cohere) have their own list of possible values?